p4proto/kctrl/switchsai/README.md

```diff
@@ -0,0 +1,98 @@
+
+For each feature, a module is initialized in sai_initialize and the pointer is returned to invoke SwitchSAI calls from the above layer. In our case, switchLink queries for each module pointer and call SwitchSAI accordingly.
+```
+Eg:
+sai_fdb_api_t fdb_api = {.create_fdb_entry = sai_create_fdb_entry,
+                                       .remove_fdb_entry = sai_remove_fdb_entry};
+
+sai_status_t sai_fdb_initialize(sai_api_service_t *sai_api_service) {
+  sai_api_service->fdb_api = fdb_api;
+  return SAI_STATUS_SUCCESS;
+}
+```
+Each SwitchLink message is constructed with attributes defined by opencompute/SAI and appropriate calls to the SwitchSAI is made. The arguments used for the SwitchSAI will be defined by the opencompute/SAI.
+Here are each SwitchSAI calls made from SwitchLink and associated SwitchAPI calls.
+
+## VRF <a name="vrf"></a>
+
+**create_virtual_router:** Takes VRF attributes as input, creates a VRF in SwitchAPI with switch_api_vrf_create and returns a handler corresponding to VRF.
+**remove_virtual_router:** Takes VRF handler as an input, deletes a VRF in SwitchAPI with switch_api_vrf_delete.
+
+## RIF <a name="rif"></a>
+
+* **create_router_interface:** Takes RIF attributes as input,
+  * Creates RMAC group with switch_api_router_mac_group_create, returns the handler to Switchlink layer for this RMAC group to be saved in RIF SwitchAPI structure
+  * Add RMAC entries for the group with switch_api_router_mac_add, returns RMAC handler to Switchlink layer for each entry to be added to the list in a RMAC group.
+  * Creates RIF interface with switch_api_rif_create, returns handler called as interface handler to Switchlink layer which is used for cross reference like neighbor, nexthop.
+* **remove_router_interface:** Takes RIF handler as input,
+  * Gets the RMAC group handler from the RIF handler (separate call to SwitchAPI switch_api_device_default_rmac_handle_get)
+  * Using RMAC group handler, loop through list of RMAC for the port and delete each RMAC and finally the RMAC group.
+  * Deletes RIF with switch_api_rif_delete.
+
+## Tunnel <a name="tunnel"></a>
+
+* **create_tunnel:** Takes tunnel attributes as input, creates tunnel entry in SwitchAPI with switch_api_tunnel_create, and returns tunnel handler to the caller.
+* **remove_tunnel:** Takes tunnel handler as an input, deletes tunnel entry in SwitchAPI with switch_api_tunnel_delete.
+* **create_tunnel_term_table_entry:** Takes tunnel terminate attributes as input, creates tunnel term entry in SwitchAPI with switch_api_tunnel_term_create and returns the handler to the caller.
+* **remove_tunnel_term_table_entry:** Takes tunnel termination handler as an input, deletes tunnel termination entry in SwitchAPI with switch_api_tunnel_term_delete.
+
+## FDB <a name="fdb"></a>
+
+FDB entry is populated from 2 sources, one is from SwitchLink→SwitchSAI→SwitchAPI and other is from OVS Mac Learning→ SwitchAPI. Here we talk only about FDB entry from SwitchLink.
```

Comment:
somehow these are shown in a different way while we view it on browser.
https://github.com/ipdk-io/ovs/blob/linux_networking_documentation/p4proto/kctrl/switchapi/README.md#fdb-